### PR TITLE
created first import for ui custom elements on dialog open

### PR DESCRIPTION
**Pull request by diliger**
added another custom element
starcounter-debug-aid-ui

and moved all ui and imports to there

import the element once on dialog-open

### CODE REVIEW

**Comment by warpech** [comment]
Implements: #42

File: src/starcounter-debug-aid-ui.html [hunk 1]
```diff
@@ -0,0 +1,111 @@
+﻿<link rel="import" href="puppet-js-view-model-tree.html" />
```
Comment (warpech) [inline src/starcounter-debug-aid-ui.html]: There should be a import of "polymer.html" here as well. Every web component needs to import all of its dependencies, not assume that they are imported already.

File: src/starcounter-debug-aid.html [hunk 2]
```diff
@@ -173,6 +81,11 @@
                 //this.$.errorDialog.reset();
             },
             openDialog: function () {
+                if (!Polymer.isInstance(this.$.ui)) {
+                    Polymer.Base.importHref("src/starcounter-debug-aid-ui.html", () => {
+                        
+                    });
+                }
```
Comment (warpech) [inline src/starcounter-debug-aid.html]: Is the empty callback needed?

File: src/starcounter-debug-aid.html [hunk 3]
```diff
@@ -173,6 +81,11 @@
                 //this.$.errorDialog.reset();
             },
             openDialog: function () {
+                if (!Polymer.isInstance(this.$.ui)) {
+                    Polymer.Base.importHref("src/starcounter-debug-aid-ui.html", () => {
```
Comment (warpech) [inline src/starcounter-debug-aid.html]: Path `src/starcounter-debug-aid-ui.html` fails to work in a Starcounter app. Polymer resolves it to an absolute link (`/src/starcounter-debug-aid-ui.html`). 

It needs to be relative to `starcounter-debug-aid.html` or it won't load in a Starcounter app, which has the external dependencies usually under `/sys/`.

2016-12-09 15_41_08-launcher

**Comment by warpech** [comment]
Also, I think that I made a mistake about which files to remove. Removing `puppet-js-listener.js` causes the Patch history in the "Patches" tab to be not recorded until `starcounter-debug-aid` is opened. Obviously we want the history to be observed from the beginning for the debugging purposes.

File: src/starcounter-debug-aid.html (hunk 3 above)
Comment (warpech) [inline src/starcounter-debug-aid.html]: I think you need to investigate deeper. Maybe this is a bug in Polymer or Polyserve. Can you explain if Polymer makes any transformations on URLs provided to `importHref`? I have looked in the Polymer issue board and cannot find issues reported to it, but maybe there should be.

Comment (diliger) [inline src/starcounter-debug-aid.html]: Maybe I am missing something about how polyserve works...

Here is what I see:
- if I place `<link rel="import" href="whatever.html" />` it results to `http://127.0.0.1:8000/components/starcounter-debug-aid/src/whatever.html`
- if I do `Polymer.Base.importHref("whatever.html");` it results to `http://127.0.0.1:8000/components/starcounter-debug-aid/whatever.html` **/src** part is missing.

**Comment by warpech** [comment]
@diliger Thanks for the fixes, I will review tomorrow. Have you also tested with a Starcounter app? (by substituting starcounter-debug-aid that comes in Starcounter installation).

**Comment by warpech** [comment]
Still does not work in Starcounter. The reason seems obvious: Polymer `baseHref` resolves the URI relatively to the main document, not to the import.

It should be fixed by determining the absolute URL before the line https://github.com/Starcounter/starcounter-debug-aid/blob/813f1a6306cb2f3f812ce3c956fe238c1e9c030b/src/starcounter-debug-aid.html#L86 in and passing that absolute URL to `baseHref`.

2016-12-20 12_02_55-launcher

**Comment by diliger** [comment]
maybe that's a simple question, but how can i determine  
> the absolute URL

for imports?

**Comment by warpech** [comment]
Try what these guys are doing: https://github.com/navsreks/nr-state-picker/blob/98da6c27bd6d5447714f8255c15f5b1a37055296/bower_components/vaadin-core-elements/demo/common.html

**Comment by warpech** [comment]
I have added the solution to determining the absolute URL. Merging as done.

Thanks @diliger

File: src/starcounter-debug-aid.html [hunk 4]
```diff
@@ -173,6 +82,12 @@
                 //this.$.errorDialog.reset();
             },
             openDialog: function () {
+                if (!Polymer.isInstance(this.$.ui)) {
+                    var fileName = "starcounter-debug-aid.html";
+                    var link = document.querySelector('link[href$="' + fileName + '"]');
+                    var baseUrl = link.href.replace(fileName, '');
+                    Polymer.Base.importHref(baseUrl + "starcounter-debug-aid-ui.html");
```
Comment (tomalec) [inline src/starcounter-debug-aid.html]: Why don't we use just `this.importHref`? and `this.resolveUrl`?
https://www.polymer-project.org/1.0/docs/devguide/instance-methods#imports-and-urls

or even `lazy-import` https://github.com/Polymer/lazy-imports ?
//cc @diliger @warpech

Comment (warpech) [inline src/starcounter-debug-aid.html]: Thanks for the suggestion. Lazy import was mentioned twice in #42 so I suppose @diliger considered it but dismissed it for some reason. 

From what I see now, the `group` attribute would actually save us from using a middle element `starcounter-debug-aid-ui.html`. But, at the cost of importing `lazy-imports.html` ;)

On the other hand it looks that `this. importHref` does not have the base URL problem of `Polymer.Base.importHref`.

Comment (warpech) [inline src/starcounter-debug-aid.html]: I created an issue for this https://github.com/Starcounter/starcounter-debug-aid/issues/44

**Pull request merged**